File: frontend/src/api/workflows.ts
import { request } from './client';

export type WorkflowStatus = 'draft' | 'active' | 'paused' | 'archived';
export type WorkflowExecutionStatus = 'pending' | 'running' | 'waiting' | 'completed' | 'failed' | 'cancelled';
export type WorkflowNodeStatus = 'pending' | 'running' | 'waiting' | 'completed' | 'failed' | 'skipped' | 'cancelled';

export type WorkflowNodeType =
  | 'start'
  | 'end'
  | 'agent'
  | 'rag'
  | 'graph'
  | 'memory'
  | 'mcp_tool'
  | 'mcp_resource'
  | 'mcp_prompt'
  | 'local_tool'
  | 'condition'
  | 'human_approval'
  | 'transform';

export interface WorkflowNode {
  id: string;
  workflow_id: string;
  node_key: string;
  node_type: WorkflowNodeType;
  name: string;
  config: Record<string, any>;
  position: { x: number; y: number };
  is_enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface WorkflowEdge {
  id: string;
  workflow_id: string;
  source_node_id: string;
  target_node_id: string;
  condition?: Record<string, any> | null;
  priority: number;
  created_at: string;
}

export interface WorkflowVariable {
  id: string;
  workflow_id: string;
  name: string;
  value?: string | null;
  value_type: string;
  is_secret: boolean;
  created_at: string;
  updated_at: string;
}

export interface WorkflowSummary {
  id: string;
  user_id: string;
  workspace_id: string;
  name: string;
  description?: string | null;
  status: WorkflowStatus;
  version: number;
  is_active: boolean;
  node_count: number;
  edge_count: number;
  created_at: string;
  updated_at: string;
}

export interface WorkflowDetail extends WorkflowSummary {
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  variables: WorkflowVariable[];
}

export interface WorkflowValidationItem {
  code: string;
  message: string;
  node_key?: string | null;
  edge_id?: string | null;
}

export interface WorkflowValidationResult {
  valid: boolean;
  errors: WorkflowValidationItem[];
  warnings: WorkflowValidationItem[];
}

export interface WorkflowExecutionSummary {
  id: string;
  workflow_id: string;
  workflow_version: number;
  user_id: string;
  workspace_id: string;
  status: WorkflowExecutionStatus;
  input_data: Record<string, any>;
  output_data?: Record<string, any> | null;
  error?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
}

export interface WorkflowExecutionNodeRecord {
  id: string;
  execution_id: string;
  node_id?: string | null;
  node_key: string;
  status: WorkflowNodeStatus;
  input_data?: Record<string, any> | null;
  output_data?: Record<string, any> | null;
  error?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface WorkflowExecutionDetail extends WorkflowExecutionSummary {
  execution_nodes: WorkflowExecutionNodeRecord[];
}

export const getWorkflows = async (params?: { limit?: number; offset?: number; status?: string }): Promise<{ workflows: WorkflowSummary[]; total: number }> => {
  const query = new URLSearchParams();
  if (params?.limit) query.set('limit', String(params.limit));
  if (params?.offset) query.set('offset', String(params.offset));
  if (params?.status) query.set('status', params.status);
  const qStr = query.toString();
  return request<{ workflows: WorkflowSummary[]; total: number }>(`/workflows${qStr ? `?${qStr}` : ''}`, { method: 'GET' });
};

export const getWorkflow = async (id: string): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}`, { method: 'GET' });
};

export const createWorkflow = async (payload: {
  name: string;
  description?: string;
  nodes?: any[];
  edges?: any[];
  variables?: any[];
}): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>('/workflows', {
    method: 'POST',
    body: JSON.stringify(payload)
  });
};

export const updateWorkflow = async (
  id: string,
  payload: {
    name?: string;
    description?: string;
    status?: WorkflowStatus;
    is_active?: boolean;
    nodes?: any[];
    edges?: any[];
    variables?: any[];
  }
): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}`, {
    method: 'PUT',
    body: JSON.stringify(payload)
  });
};

export const getWorkflowDefinition = async (id: string): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}/definition`, { method: 'GET' });
};

export const updateWorkflowDefinition = async (
  id: string,
  payload: {
    expected_version: number;
    name?: string;
    description?: string;
    nodes: any[];
    edges: any[];
    variables?: any[];
  }
): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}/definition`, {
    method: 'PUT',
    body: JSON.stringify(payload)
  });
};

export const cloneWorkflow = async (id: string, name?: string): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}/clone`, {
    method: 'POST',
    body: JSON.stringify({ name })
  });
};

export const deleteWorkflow = async (id: string): Promise<void> => {
  return request<void>(`/workflows/${id}`, { method: 'DELETE' });
};

export const validateWorkflow = async (id: string): Promise<WorkflowValidationResult> => {
  return request<WorkflowValidationResult>(`/workflows/${id}/validate`, { method: 'POST' });
};

export const activateWorkflow = async (id: string): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}/activate`, { method: 'POST' });
};

export const pauseWorkflow = async (id: string): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}/pause`, { method: 'POST' });
};

export const archiveWorkflow = async (id: string): Promise<WorkflowDetail> => {
  return request<WorkflowDetail>(`/workflows/${id}/archive`, { method: 'POST' });
};

export const executeWorkflow = async (id: string, inputData: Record<string, any> = {}): Promise<WorkflowExecutionDetail> => {
  return request<WorkflowExecutionDetail>(`/workflows/${id}/execute`, {
    method: 'POST',
    body: JSON.stringify({ input_data: inputData })
  });
};

export const getWorkflowExecutions = async (id: string, params?: { limit?: number; offset?: number }): Promise<WorkflowExecutionSummary[]> => {
  const query = new URLSearchParams();
  if (params?.limit) query.set('limit', String(params.limit));
  if (params?.offset) query.set('offset', String(params.offset));
  const qStr = query.toString();
  return request<WorkflowExecutionSummary[]>(`/workflows/${id}/executions${qStr ? `?${qStr}` : ''}`, { method: 'GET' });
};

export const getWorkflowExecution = async (executionId: string): Promise<WorkflowExecutionDetail> => {
  return request<WorkflowExecutionDetail>(`/workflows/executions/${executionId}`, { method: 'GET' });
};

export const cancelWorkflowExecution = async (executionId: string): Promise<WorkflowExecutionSummary> => {
  return request<WorkflowExecutionSummary>(`/workflows/executions/${executionId}/cancel`, { method: 'POST' });
};

export const approveWorkflowExecution = async (
  executionId: string,
  approved: boolean = true,
  comments?: string
): Promise<WorkflowExecutionSummary> => {
  return request<WorkflowExecutionSummary>(`/workflows/executions/${executionId}/approve`, {
    method: 'POST',
    body: JSON.stringify({ approved, comments })
  });
};

export interface WorkflowApproval {
  id: string;
  execution_id: string;
  workflow_id: string;
  workflow_node_id?: string | null;
  workspace_id: string;
  node_key: string;
  requested_by: string;
  assigned_roles: string[];
  assigned_users: string[];
  status: string;
  policy: string;
  required_count: number;
  requester_can_approve: boolean;
  title: string;
  message?: string | null;
  timeout_seconds: number;
  expires_at?: string | null;
  decided_by?: string | null;
  decided_at?: string | null;
  decision?: string | null;
  reason?: string | null;
  decision_history: any[];
  created_at: string;
  updated_at: string;
}

export const getWorkflowApprovals = async (params?: {
  status?: string;
  limit?: number;
  offset?: number;
}): Promise<{ approvals: WorkflowApproval[]; total: number }> => {
  const query = new URLSearchParams();
  if (params?.status) query.set('status', params.status);
  if (params?.limit) query.set('limit', String(params.limit));
  if (params?.offset) query.set('offset', String(params.offset));
  const qStr = query.toString();
  return request<{ approvals: WorkflowApproval[]; total: number }>(`/workflows/approvals${qStr ? `?${qStr}` : ''}`, {
    method: 'GET'
  });
};

export const getWorkflowApproval = async (approvalId: string): Promise<WorkflowApproval> => {
  return request<WorkflowApproval>(`/workflows/approvals/${approvalId}`, { method: 'GET' });
};

export const approveWorkflowApproval = async (
  approvalId: string,
  reason?: string
): Promise<WorkflowApproval> => {
  return request<WorkflowApproval>(`/workflows/approvals/${approvalId}/approve`, {
    method: 'POST',
    body: JSON.stringify({ decision: 'approved', reason })
  });
};

export const rejectWorkflowApproval = async (
  approvalId: string,
  reason?: string
): Promise<WorkflowApproval> => {
  return request<WorkflowApproval>(`/workflows/approvals/${approvalId}/reject`, {
    method: 'POST',
    body: JSON.stringify({ decision: 'rejected', reason })
  });
};

File: backend/app/api/v1/endpoints/workflows.py
import uuid
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from loguru import logger

from app.database.session import get_db
from app.api.dependencies import get_current_user, check_rate_limit
from app.models.user import User
from app.api.v1.endpoints.documents import resolve_workspace_id

from app.models.workflow import (
    WorkflowStatus,
    WorkflowExecutionStatus
)
from app.schemas.workflow import (
    WorkflowCreate,
    WorkflowUpdate,
    WorkflowResponse,
    WorkflowDetailResponse,
    WorkflowListResponse,
    WorkflowValidationResult,
    WorkflowExecutionCreate,
    WorkflowExecutionResponse,
    WorkflowExecutionDetailResponse,
    WorkflowNodeResponse,
    WorkflowEdgeResponse,
    WorkflowVariableResponse,
    WorkflowExecutionNodeResponse,
    WorkflowDefinitionUpdate,
    WorkflowCloneRequest,
    WorkflowApproveRequest
)
from app.services.workflow import (
    WorkflowService,
    VersionConflictError,
    WorkflowArchivedError
)
from app.services.workflow_execution import WorkflowExecutionService

router = APIRouter(prefix="/workflows", tags=["Workflows Engine"])

@router.post("", response_model=WorkflowDetailResponse, status_code=status.HTTP_201_CREATED, dependencies=[Depends(check_rate_limit)])
async def create_workflow(
    payload: WorkflowCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Creates a new workflow definition with optional initial nodes, edges, and variables.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow = service.create_workflow(current_user.id, workspace_id, payload)
    return workflow

@router.get("", response_model=WorkflowListResponse, dependencies=[Depends(check_rate_limit)])
async def list_workflows(
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    status: Optional[str] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Lists all workflows in the authenticated workspace.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflows, total = service.list_workflows(current_user.id, workspace_id, limit, offset, status)

    items = [
        WorkflowResponse(
            id=w.id,
            user_id=w.user_id,
            workspace_id=w.workspace_id,
            name=w.name,
            description=w.description,
            status=w.status,
            version=w.version,
            is_active=w.is_active,
            node_count=len([n for n in w.nodes if not n.deleted_at]),
            edge_count=len([e for e in w.edges if not e.deleted_at]),
            created_at=w.created_at,
            updated_at=w.updated_at
        )
        for w in workflows
    ]

    return WorkflowListResponse(
        workflows=items,
        total=total,
        limit=limit,
        offset=offset
    )

@router.get("/{workflow_id}", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def get_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Retrieves full details of a workflow, including nodes, edges, and variables.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow = service.get_workflow(current_user.id, workspace_id, workflow_id)
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return workflow

@router.put("/{workflow_id}", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def update_workflow(
    workflow_id: uuid.UUID,
    payload: WorkflowUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Updates workflow metadata or structure (increments version if structural changes are made).
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow = service.update_workflow(current_user.id, workspace_id, workflow_id, payload)
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return workflow

@router.get("/{workflow_id}/definition", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def get_workflow_definition(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Retrieves the complete graph definition of a workflow for visual canvas loading.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow = service.get_workflow_definition(current_user.id, workspace_id, workflow_id)
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return workflow

@router.put("/{workflow_id}/definition", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def update_workflow_definition(
    workflow_id: uuid.UUID,
    payload: WorkflowDefinitionUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Atomically updates the complete visual workflow graph definition with optimistic concurrency protection.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    try:
        workflow = service.update_workflow_definition(current_user.id, workspace_id, workflow_id, payload)
        if not workflow:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Workflow '{workflow_id}' not found in active workspace."
            )
        return workflow
    except VersionConflictError as e:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(e)
        )
    except WorkflowArchivedError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )

@router.post("/{workflow_id}/clone", response_model=WorkflowDetailResponse, status_code=status.HTTP_201_CREATED, dependencies=[Depends(check_rate_limit)])
async def clone_workflow(
    workflow_id: uuid.UUID,
    payload: Optional[WorkflowCloneRequest] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Clones an existing workflow into a new draft workflow.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    cloned = service.clone_workflow(
        current_user.id,
        workspace_id,
        workflow_id,
        clone_name=payload.name if payload else None
    )
    if not cloned:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return cloned

@router.delete("/{workflow_id}", status_code=status.HTTP_204_NO_CONTENT, dependencies=[Depends(check_rate_limit)])
async def delete_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Soft deletes a workflow and cleans up active references.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    deleted = service.delete_workflow(current_user.id, workspace_id, workflow_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return None

@router.post("/{workflow_id}/validate", response_model=WorkflowValidationResult, dependencies=[Depends(check_rate_limit)])
async def validate_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Runs deterministic DAG validation against the workflow structure.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    result = service.validate_workflow(current_user.id, workspace_id, workflow_id)
    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return result

@router.post("/{workflow_id}/activate", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def activate_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Validates and activates a workflow for production execution.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow, validation = service.activate_workflow(current_user.id, workspace_id, workflow_id)
    if not workflow:
        if validation and not validation.valid:
            err_msgs = "; ".join([e.message for e in validation.errors])
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Workflow validation failed: {err_msgs}"
            )
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return workflow

@router.post("/{workflow_id}/pause", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def pause_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Pauses an active workflow.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow = service.pause_workflow(current_user.id, workspace_id, workflow_id)
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return workflow

@router.post("/{workflow_id}/archive", response_model=WorkflowDetailResponse, dependencies=[Depends(check_rate_limit)])
async def archive_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Archives a workflow.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    service = WorkflowService(db)
    workflow = service.archive_workflow(current_user.id, workspace_id, workflow_id)
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow '{workflow_id}' not found in active workspace."
        )
    return workflow

@router.post("/{workflow_id}/execute", response_model=WorkflowExecutionDetailResponse, dependencies=[Depends(check_rate_limit)])
async def execute_workflow(
    workflow_id: uuid.UUID,
    payload: WorkflowExecutionCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Executes a workflow with the provided input parameters.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    exec_service = WorkflowExecutionService(db)
    try:
        execution = exec_service.execute_workflow(
            user_id=current_user.id,
            workspace_id=workspace_id,
            workflow_id=workflow_id,
            input_data=payload.input_data
        )
        return execution
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        logger.error(f"Execution error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Workflow execution failed due to an internal error."
        )

@router.get("/{workflow_id}/executions", response_model=List[WorkflowExecutionResponse], dependencies=[Depends(check_rate_limit)])
async def list_workflow_executions(
    workflow_id: uuid.UUID,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Lists execution history for a given workflow.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    exec_service = WorkflowExecutionService(db)
    executions, _ = exec_service.list_executions(current_user.id, workspace_id, workflow_id, limit, offset)
    return executions

@router.get("/executions/{execution_id}", response_model=WorkflowExecutionDetailResponse, dependencies=[Depends(check_rate_limit)])
async def get_workflow_execution(
    execution_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Retrieves execution trace details for a specific workflow execution.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    exec_service = WorkflowExecutionService(db)
    execution = exec_service.get_execution(current_user.id, workspace_id, execution_id)
    if not execution:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow execution '{execution_id}' not found."
        )
    return execution

@router.post("/executions/{execution_id}/cancel", response_model=WorkflowExecutionResponse, dependencies=[Depends(check_rate_limit)])
async def cancel_workflow_execution(
    execution_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Cancels an active or queued workflow execution.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    exec_service = WorkflowExecutionService(db)
    execution = exec_service.cancel_execution(current_user.id, workspace_id, execution_id)
    if not execution:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workflow execution '{execution_id}' not found."
        )
    return execution

@router.post("/executions/{execution_id}/approve", response_model=WorkflowExecutionResponse, dependencies=[Depends(check_rate_limit)])
async def approve_workflow_execution(
    execution_id: uuid.UUID,
    payload: WorkflowApproveRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Approves or rejects a workflow execution paused in WAITING_APPROVAL status.
    """
    workspace_id = resolve_workspace_id(current_user, db)
    exec_service = WorkflowExecutionService(db)
    try:
        execution = exec_service.approve_execution(
            current_user.id,
            workspace_id,
            execution_id,
            approved=payload.approved
        )
        return execution
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
